File: app/schemas/activity.py
"""Recent activity: one recency-ordered stream; status is a label and never affects order."""

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

ActivityType = Literal["mission", "report", "evidence"]


class ActivityItem(BaseModel):
    type: ActivityType
    id: UUID
    title: str
    subtitle: str | None = None
    status: str | None = None
    occurred_at: datetime
    href: str
    new: bool


class ActivityPage(BaseModel):
    generated_at: datetime
    refresh_seconds: int = 30
    page: int
    page_size: int
    total: int
    new_total: int
    items: list[ActivityItem]


class ActivitySummary(BaseModel):
    generated_at: datetime
    new_total: int
    by_type: dict[str, int]


class ViewedItem(BaseModel):
    type: ActivityType
    id: UUID
    occurred_at: datetime


class MarkViewedRequest(BaseModel):
    items: list[ViewedItem] = Field(min_length=1, max_length=200)


class MarkViewedResponse(BaseModel):
    viewed: int
    new_total: int

File: frontend/src/lib/api/activity.ts
import { httpClient } from "@/lib/api/http";

export type ActivityType = "mission" | "report" | "evidence";
export interface ActivityItem {
  type: ActivityType;
  id: string;
  title: string;
  subtitle: string | null;
  status: string | null;
  occurred_at: string;
  href: string;
  new: boolean;
}
export interface ActivityPage {
  generated_at: string;
  refresh_seconds: number;
  page: number;
  page_size: number;
  total: number;
  new_total: number;
  items: ActivityItem[];
}
export interface ActivitySummary { generated_at: string; new_total: number; by_type: Record<string, number> }
export interface ViewedItem { type: ActivityType; id: string; occurred_at: string }

export const activityApi = {
  list: (params: { page?: number; page_size?: number } = {}) => httpClient.get<ActivityPage>("/activity", { params }),
  summary: () => httpClient.get<ActivitySummary>("/activity/summary"),
  markViewed: (items: ViewedItem[]) => httpClient.put<{ viewed: number; new_total: number }>("/activity/viewed", { items }),
};

export const ACTIVITY_SECTIONS: Record<ActivityType, { label: string; href: string }> = {
  mission: { label: "Missions", href: "/missions" },
  report: { label: "Reports", href: "/reports" },
  evidence: { label: "Evidence", href: "/evidence" },
};

export function newBadgeName(label: string, count: number) {
  return `${label}, ${count.toLocaleString()} new`;
}

/** The moment a mission last happened, matching the server's activity order. */
export function missionOccurredAt(mission: { status: string; completed_at?: string | null; started_at?: string | null; queued_at?: string | null; created_at?: string; updated_at: string }) {
  if (mission.status === "completed") return mission.completed_at ?? mission.updated_at;
  if (mission.status === "in_progress") return mission.started_at ?? mission.updated_at;
  if (mission.status === "queued") return mission.queued_at ?? mission.created_at ?? mission.updated_at;
  return mission.updated_at;
}
